In [33]:
import time

from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.pipeline import make_pipeline

from utils import open_mnist_or_download_if_missing, shuffle_split

%matplotlib inline

RS = 33

In [3]:
mnist = open_mnist_or_download_if_missing()

Using local mnist


In [4]:
X_train, y_train, X_test, y_test = shuffle_split(mnist["data"], mnist["target"], 60000, RS)

In [31]:
def time_action(func, *args, **kwargs):
    start_time = time.time()
    func(*args, **kwargs)
    return time.time() - start_time

def time_fit(clf, X, y):
    return time_action(clf.fit, X, y)

In [10]:
clf = RandomForestClassifier()
time_fit(clf, X_train, y_train)

25.50989294052124

In [12]:
accuracy_score(y_test, clf.predict(X_test))

0.9669

So random forest classifier took 25 seconds to fit with resulting accuracy of 0.9669

Now let's see how it will do with PCA with explained variance ratio of 95%

In [48]:
pipeline = make_pipeline(PCA(n_components=0.95), RandomForestClassifier())
time_fit(pipeline, X_train, y_train)

73.97757411003113

In [51]:
accuracy_score(y_test, pipeline.predict(X_test))

0.9446

So, it is way slover, took more then 73 seconds and accuracy is a bit lower (expected I guess) - 0.9446